presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -1823,16 +1834,25 @@ public void updateTableWriteId(HiveIdentity identity, String dbName, String tabl
     }
 
     @Override
-    public void alterPartitions(HiveIdentity identity, String dbName, String tableName, List<Partition> partitions, long writeId)
+    public void alterPartitions(HiveIdentity identity, String dbName, String tableName, List<Partition> partitions, EnvironmentContext context, long writeId)
     {
         checkArgument(writeId > 0, "writeId should be a positive integer, but was %s", writeId);
         try {
             retry()
                     .stopOnIllegalExceptions()
                     .run("alterPartitions", stats.getAlterPartitions().wrap(() -> {
-                        try (ThriftMetastoreClient metastoreClient = createMetastoreClient(identity)) {
-                            metastoreClient.alterPartitions(dbName, tableName, partitions, writeId);
-                        }
+                        alternativeCall(
+                                () -> createMetastoreClient(identity),
+                                exception -> !isUnknownMethodExceptionalResponse(exception),
+                                chosenAlterPartitionsAlternative,
+                                client -> {
+                                    client.alterPartitions(dbName, tableName, partitions, writeId);
+                                    return null;
+                                },
+                                client -> {
+                                    client.alterPartitionsWithEnvironmentContext(dbName, tableName, partitions, context);
```

Comment:
Would it be correct to use this call only (avoid need for  alternativeCall here)

cc @djsstarburst

Reply:
It's a great question, @findepi.  I _believe_ that there must be some value in the final update of the Hive metastore writeId at the end of the transaction.  I _think_ that value is identifying transactions that never completed in spite of having written delta and delete_delta files.  I spent a bit of time trying to figure out what the final writeId update accomplished, but the Hive codebase is challenging.

I'm not aware of any Hive ACID concurrency tests that would help us figure this out.

Reply:
@djsstarburst Wouldn't the combination of allocateWriteId and commitTransaction achieve similar semantics for transactions that could not be completed/committed in spite of having written the delta and delete_delta files? I mean through allocateWriteId the metastore is being made aware of the [intended]writeId for the transaction. In case the transaction is not committed in the metastore(commitTransaction) after having written delta and delete_delta files, the metastore is still aware of such files through the relation between the transactionId and the allocatedWriteId. 

After testing the first  commit with HMS 3.1.2 on EMR I found that the updateWriteId was also calling alter_table_req. So I made the alternative call change for that part as well and then I was able to manually test it successfully. From these changes I get a feeling that the updateWriteId and the alter table in the end to update the writeId/validWriteIds might not be needed. But yeah, we need a way to be certain, otherwise even the alternative change would have issues because of not updating the writeId.

Reply:
That's great, @abhishekkhanna1!